File: python/day20/part1.py
#!/usr/bin/env python3

class Node:
    def __init__(self, id, val, prev=None):
        self.id = id
        self.val = val
        self.prev = prev
        self.next = None


def parse(input_path, decryption_key):
    with open(input_path, 'r') as f:
        count = 0
        start = None
        curr = None
        # build the linked-list
        for val in [decryption_key * int(line) for line in f.readlines()]:
            curr = Node(count, val, curr)
            count += 1
            if start is None:
                start = curr
            else:
                curr.prev.next = curr  # bidirectional-link
        # close the loop
        curr.next = start
        start.prev = curr
        return count, start


def solve_part1(parsed, loops):
    count, curr = parsed
    for k in range(loops):
        for i in range(count):
            # find the node to move
            while curr.id != i:
                curr = curr.next
            shift = curr.val % (count - 1)
            if shift == 0:
                continue
            if shift > (count - 1)//2 :
                shift -= (count - 1)
            # find the position it should go to
            next_pos = curr
            if shift > 0:
                for _ in range(shift + 1):
                    next_pos = next_pos.next
            else:
                for _ in range(-shift):
                    next_pos = next_pos.prev
            # re-arrange the links between curr and its next position
            curr.prev.next = curr.next
            curr.next.prev = curr.prev
            curr.prev = next_pos.prev
            curr.prev.next = curr
            curr.next = next_pos
            next_pos.prev = curr

    # done with all permutations, calculate the result
    while curr.val != 0:
        curr = curr.next
    res = 0
    for i in range(3):
        for _ in range(1000):
            curr = curr.next
        res += curr.val
    return res


if __name__ == "__main__":
    parsed = parse("data.txt", 1)
    print('Part 1 :', solve_part1(parsed, 1))

File: c++/day20/part1.cpp
#include "part1.h"

using namespace std;

namespace Day20 {

    shared_ptr<Node> parse(const string& filename) {
        shared_ptr<Node> firstNode { nullptr };
        shared_ptr<Node> currNode  { nullptr };
        int nodeId { 0 };
        for (const auto line : getFileLines(filename)) {
            auto node = make_shared<Node>();
            node->val = stoll(line);
            node->nodeId = nodeId++;
            if (firstNode == nullptr) {
                firstNode = node;
                currNode = node;
            } else {
                node->prev = currNode;
                currNode->next = node;
                currNode = node;
            }
        }
        currNode->next = firstNode;
        firstNode->prev = currNode;
        return firstNode;
    }


    namespace Part1 {

        void mix(const shared_ptr<Node>& firstNode, const int maxNodeId) {
            auto curr = firstNode;
            for (int nodeId = 0; nodeId <= maxNodeId; ++nodeId) {
                // set the cursor to the node to move
                while (curr->nodeId != nodeId) {
                    curr = curr->next;
                }
                // get the number of elements to move the node forward
                const int shift = (curr->val % maxNodeId) + (curr-> val < 0 ? maxNodeId : 0);
                if (shift == 0) {
                    continue;
                }
                // find the node right before where this node should move
                auto target = curr;
                for (int i = 0; i < shift; ++i) {
                    target = target->next;
                }
                // insert the node to its target position
                curr->prev.lock()->next = curr->next;
                curr->next->prev = curr->prev;
                target->next->prev = curr;
                curr->next = target->next;
                target->next = curr;
                curr->prev = target;
            }
        }


        long long groveCoordinate(const shared_ptr<Node>& firstNode) {
            auto curr = firstNode;
            // move the current pointer to the node with value zero
            while (curr->val != 0LL) {
                curr = curr->next;
            }
            // calculate the sum of the 1000th, 2000th and 3000th elements
            long long result = 0;
            for (int i = 0; i < 3; ++i) {
                for (int k = 0; k < 1000; ++k) {
                    curr = curr->next;
                }
                result += curr->val;
            }
            return result;
        }


        long long solve(const shared_ptr<Node>& firstNode) {
            const int maxNodeId = firstNode->prev.lock()->nodeId;
            mix(firstNode, maxNodeId);
            return groveCoordinate(firstNode);
        }

    }
}
